internal/servers/common/handler.go

```diff
@@ -0,0 +1,223 @@
+// optional interfaces: http.Flusher, http.Pusher and http.Hijacker.  It return
+// an http.ResponseWriter that satisfies those optional interfaces.
+//
+// See: https://medium.com/@cep21/interface-wrapping-method-erasure-c523b3549912
+func WrapWithOptionals(with *writerWrapper, wrap http.ResponseWriter) (http.ResponseWriter, error) {
+	const op = "common.WrapWithOptionals"
+	if with == nil {
+		return nil, errors.New(errors.InvalidParameter, op, "missing writer wrapper")
+	}
+	if wrap == nil {
+		return nil, errors.New(errors.InvalidParameter, op, "missing response writer")
+	}
+	flusher, _ := wrap.(http.Flusher)
+	pusher, _ := wrap.(http.Pusher)
+	hijacker, _ := wrap.(http.Hijacker)
+	switch {
+	case flusher == nil && pusher == nil && hijacker == nil:
+		return with, nil
+	case flusher != nil && pusher == nil && hijacker == nil:
+		return struct {
+			*writerWrapper
+			http.Flusher
+		}{with, flusher}, nil
+	case flusher == nil && pusher != nil && hijacker == nil:
+		return struct {
+			*writerWrapper
+			http.Pusher
+		}{with, pusher}, nil
+	case flusher == nil && pusher == nil && hijacker != nil:
+		return struct {
+			*writerWrapper
+			http.Hijacker
+		}{with, hijacker}, nil
+	case flusher != nil && pusher != nil && hijacker == nil:
+		return struct {
+			*writerWrapper
+			http.Flusher
+			http.Pusher
+		}{with, flusher, pusher}, nil
+	case flusher != nil && pusher == nil && hijacker != nil:
+		return struct {
+			*writerWrapper
+			http.Flusher
+			http.Hijacker
+		}{with, flusher, hijacker}, nil
+	case flusher == nil && pusher != nil && hijacker != nil:
+		return struct {
+			*writerWrapper
+			http.Pusher
+			http.Hijacker
+		}{with, pusher, hijacker}, nil
+	default:
+		return struct {
+			*writerWrapper
+			http.Flusher
+			http.Pusher
+			http.Hijacker
+		}{with, flusher, pusher, hijacker}, nil
+	}
+}
+
+// WrapWithEventsHandler will wrap the provided http.Handler with a
+// handler that adds an Eventer to the request context and starts/flushes gated
+// events of type: observation and audit
+func WrapWithEventsHandler(h http.Handler, e *event.Eventer, logger hclog.Logger, kms *kms.Kms) (http.Handler, error) {
+	const op = "common.WrapWithEventsHandler"
+	if h == nil {
+		return nil, errors.New(errors.InvalidParameter, op, "missing handler")
+	}
+	if e == nil {
+		return nil, errors.New(errors.InvalidParameter, op, "missing eventer")
+	}
+	if logger == nil {
+		return nil, errors.New(errors.InvalidParameter, op, "missing logger")
+	}
+	if kms == nil {
+		return nil, errors.New(errors.InvalidParameter, op, "missing kms")
+	}
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		publicId, _, _ := auth.GetTokenFromRequest(logger, kms, r)
+
+		var err error
+		info := &event.RequestInfo{
+			Id:       generatedTraceId(ctx),
+			PublicId: publicId,
+			Method:   r.Method,
+			Path:     r.URL.RequestURI(),
+		}
+		ctx, err = event.NewRequestInfoContext(ctx, info)
```

Comment:
Does it make sense to combine the request info and eventer context functions? That way we're not deriving another layer of contexts.

Reply:
I kept them separate because I thought there'd be situations where you have an Eventer but no request.  Happy to discuss if that's really an edge case we shouldn't worry about.  We'd need to combine the Eventer and RequestInfo into a struct to have just one layer of context for them.